src/nvim/getchar.c

```diff
@@ -1680,6 +1680,325 @@ int char_avail(void)
+                  || mp->m_keys[1] != KS_EXTRA
+                  || mp->m_keys[2] != KE_SNR)) {
+            continue;
+          }
+
+          // If one of the typed keys cannot be remapped, skip the entry.
+          for (n = mlen; --n >= 0;) {
+            if (*s++ & (RM_NONE|RM_ABBR)) {
+              break;
+            }
+          }
+          if (n >= 0) {
+            continue;
+          }
+
+          if (keylen > typebuf.tb_len) {
+            if (!*timedout && !(mp_match != NULL && mp_match->m_nowait)) {
+              // break at a partly match
+              keylen = KEYLEN_PART_MAP;
+              break;
+            }
+          } else if (keylen > mp_match_len
+                     || (keylen == mp_match_len
+                         && mp_match != NULL
+                         && (mp_match->m_mode & LANGMAP) == 0
+                         && (mp->m_mode & LANGMAP) != 0)) {
+            // found a longer match
+            mp_match = mp;
+            mp_match_len = keylen;
+          }
+        } else {
+          // No match; may have to check for termcode at next character.
+          if (max_mlen < mlen) {
+            max_mlen = mlen;
+          }
+        }
+      }
+    }
+
+    // If no partly match found, use the longest full match.
+    if (keylen != KEYLEN_PART_MAP) {
+      mp = mp_match;
+      keylen = mp_match_len;
+    }
+  }
+
+  // Check for match with 'pastetoggle'
+  if (*p_pt != NUL && mp == NULL && (State & (INSERT|NORMAL))) {
+    bool match = typebuf_match_len(p_pt, &mlen);
+    if (match) {
+      // write chars to script file(s)
+      if (mlen > typebuf.tb_maplen) {
+        gotchars(typebuf.tb_buf + typebuf.tb_off + typebuf.tb_maplen,
+                 (size_t)(mlen - typebuf.tb_maplen));
+      }
+
+      del_typebuf(mlen, 0);  // remove the chars
+      set_option_value("paste", !p_paste, NULL, 0);
+      if (!(State & INSERT)) {
+        msg_col = 0;
+        msg_row = Rows - 1;
+        msg_clr_eos();  // clear ruler
+      }
+      status_redraw_all();
+      redraw_statuslines();
+      showmode();
+      setcursor();
+      *keylenp = keylen;
+      return map_result_retry;
+    }
+    // Need more chars for partly match.
+    if (mlen == typebuf.tb_len) {
+      keylen = KEYLEN_PART_KEY;
+    } else if (max_mlen < mlen) {
+      // no match, may have to check for termcode at next character
+      max_mlen = mlen + 1;
+    }
+  }
+
+  if ((mp == NULL || max_mlen >= mp_match_len) && keylen != KEYLEN_PART_MAP) {
+    // No matching mapping found or found a non-matching mapping that
+    // matches at least what the matching mapping matched
+    keylen = 0;
+    (void)keylen;  // suppress clang/dead assignment
+    // If there was no mapping, use the character from the typeahead
+    // buffer right here. Otherwise, use the mapping (loop around).
+    if (mp == NULL) {
+      *keylenp = keylen;
+      return map_result_get;  // got character, break for loop
```

Comment:
This comment seems to be in the wrong place. I'll considering making a PR to Vim.